Iterative panics

The following happens regularly but not always:

```
[root@zdnstest zdnstest]# head -n 1000 hosts.million | ./zdns/zdns/zdns alookup -iterative -threads 100 > results
panic: should not be able to reach here

goroutine 82 [running]:
github.com/zmap/zdns/modules/miekg.(*Lookup).iterateOnAuthorities(0xc0001e87d8, 0x10001, 0xc00001a0e7, 0x9, 0x3, 0x9e53a0, 0x0, 0x0, 0x9e53a0, 0x0, ...)
	/root/zdnstest/zdns/modules/miekg/miekg.go:771 +0x1416
github.com/zmap/zdns/modules/miekg.(*Lookup).iterativeLookup(0xc0001e87d8, 0x10001, 0xc00001a0e7, 0x9, 0xc0005abac0, 0x12, 0x3, 0xc00041ed60, 0x9, 0x9e53a0, ...)
	/root/zdnstest/zdns/modules/miekg/miekg.go:819 +0x835
github.com/zmap/zdns/modules/miekg.(*Lookup).iterateOnAuthorities(0xc0001e87d8, 0x10001, 0xc00001a0e7, 0x9, 0x2, 0x9e53a0, 0x0, 0x0, 0x9e53a0, 0x0, ...)
	/root/zdnstest/zdns/modules/miekg/miekg.go:758 +0xbe3
github.com/zmap/zdns/modules/miekg.(*Lookup).iterativeLookup(0xc0001e87d8, 0x10001, 0xc00001a0e7, 0x9, 0xc0004995e0, 0x11, 0x2, 0xc00041fc38, 0x3, 0x9e53a0, ...)
	/root/zdnstest/zdns/modules/miekg/miekg.go:819 +0x835
github.com/zmap/zdns/modules/miekg.(*Lookup).iterateOnAuthorities(0xc0001e87d8, 0x10001, 0xc00001a0e7, 0x9, 0x1, 0x9e53a0, 0x0, 0x0, 0xc000436d80, 0x8, ...)
	/root/zdnstest/zdns/modules/miekg/miekg.go:758 +0xbe3
github.com/zmap/zdns/modules/miekg.(*Lookup).iterativeLookup(0xc0001e87d8, 0x10001, 0xc00001a0e7, 0x9, 0x76a31c, 0xe, 0x1, 0x76786a, 0x1, 0x9e53a0, ...)
	/root/zdnstest/zdns/modules/miekg/miekg.go:819 +0x835
github.com/zmap/zdns/modules/miekg.(*Lookup).DoMiekgLookup(0xc0001e87d8, 0xc000010001, 0xc00001a0e7, 0x9, 0x76a31c, 0xe, 0x0, 0x0, 0x0, 0x0, ...)
	/root/zdnstest/zdns/modules/miekg/miekg.go:847 +0x2dd
github.com/zmap/zdns/modules/alookup.(*Lookup).doLookupProtocol(0xc0001e87d0, 0xc00001a0e7, 0x9, 0x76a31c, 0xe, 0x1, 0xc0008ef608, 0xc0008ef5d8, 0xc00001a0e7, 0x9, ...)
	/root/zdnstest/zdns/modules/alookup/alookup.go:62 +0x8c2
github.com/zmap/zdns/modules/alookup.(*Lookup).DoTargetedLookup(0xc0001e87d0, 0xc00001a0e7, 0x9, 0x76a31c, 0xe, 0xc000084ae0, 0x7678f6, 0x2, 0x9e6680, 0x0, ...)
	/root/zdnstest/zdns/modules/alookup/alookup.go:128 +0x186
github.com/zmap/zdns/modules/alookup.(*Lookup).DoLookup(0xc0001e87d0, 0xc00001a0e7, 0x9, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/root/zdnstest/zdns/modules/alookup/alookup.go:43 +0x6f
github.com/zmap/zdns.doLookup(0x7d9d80, 0xc00009e600, 0xc00006c900, 0xc000078120, 0xc000078180, 0xc00009e720, 0xc000098930, 0x3d, 0xc0000ed8a8, 0x41cdce)
	/root/zdnstest/zdns/lookup.go:132 +0x95d
created by github.com/zmap/zdns.DoLookups
	/root/zdnstest/zdns/lookup.go:205 +0x2e2
[root@zdnstest zdnstest]#
```

zdns is built from the current master branch. go version "go1.13.4 linux/amd64". This is being built and run on a CentOS 8 box.

--- a/modules/miekg/miekg.go
+++ b/modules/miekg/miekg.go
@@ -527,8 +527,10 @@ func (s *Lookup) iterateOnAuthorities(q Question, depth int, result Result, laye
 			// default case we continue
 			if new_status == nil && err == nil {
 				if i+1 == len(result.Authorities) {
+						s.VerboseLog((depth + 2), "--> Auth find Failed. Unknown error. Continue: ", ns_status)
 				        continue
 				} else {
+						s.VerboseLog((depth + 2), "--> Auth find Failed. Unknown error. No more authorities to try, terminating: ", ns_status)
 				        var r Result
 				        return r, trace, ns_status, err
 				}

--- a/modules/miekg/miekg.go
+++ b/modules/miekg/miekg.go
@@ -527,12 +527,12 @@ func (s *Lookup) iterateOnAuthorities(q Question, depth int, result Result, laye
 			// default case we continue
 			if new_status == nil && err == nil {
 				if i+1 == len(result.Authorities) {
-					s.VerboseLog((depth + 2), "--> Auth find Failed. Unknown error. Continue: ", ns_status)
-					continue
-				} else {
 					s.VerboseLog((depth + 2), "--> Auth find Failed. Unknown error. No more authorities to try, terminating: ", ns_status)
 					var r Result
 					return r, trace, ns_status, err
+				} else {
+					s.VerboseLog((depth + 2), "--> Auth find Failed. Unknown error. Continue: ", ns_status)
+					continue
 				}
 			} else {
 				// otherwise we hit a status we know

--- a/modules/miekg/miekg.go
+++ b/modules/miekg/miekg.go
@@ -527,12 +527,12 @@ func (s *Lookup) iterateOnAuthorities(q Question, depth int, result Result, laye
 			// default case we continue
 			if new_status == nil && err == nil {
 				if i+1 == len(result.Authorities) {
-						s.VerboseLog((depth + 2), "--> Auth find Failed. Unknown error. Continue: ", ns_status)
-				        continue
+					s.VerboseLog((depth + 2), "--> Auth find Failed. Unknown error. Continue: ", ns_status)
+					continue
 				} else {
-						s.VerboseLog((depth + 2), "--> Auth find Failed. Unknown error. No more authorities to try, terminating: ", ns_status)
-				        var r Result
-				        return r, trace, ns_status, err
+					s.VerboseLog((depth + 2), "--> Auth find Failed. Unknown error. No more authorities to try, terminating: ", ns_status)
+					var r Result
+					return r, trace, ns_status, err
 				}
 			} else {
 				// otherwise we hit a status we know

--- a/modules/miekg/miekg.go
+++ b/modules/miekg/miekg.go
@@ -526,8 +526,12 @@ func (s *Lookup) iterateOnAuthorities(q Question, depth int, result Result, laye
 			new_status, err := handleStatus(&ns_status, err)
 			// default case we continue
 			if new_status == nil && err == nil {
-				s.VerboseLog((depth + 2), "--> Auth find Failed. Unknown error. Continue: ", ns_status)
-				continue
+				if i+1 == len(result.Authorities) {
+				        continue
+				} else {
+				        var r Result
+				        return r, trace, ns_status, err
+				}
 			} else {
 				// otherwise we hit a status we know
 				var r Result

--- a/modules/mxlookup/mxlookup.go
+++ b/modules/mxlookup/mxlookup.go
@@ -69,7 +69,7 @@ func (s *Lookup) LookupIPs(name, nameServer string) (CachedAddresses, zdns.Trace
 	trace := make([]interface{}, 0)
 	// ipv4
 	if s.Factory.Factory.IPv4Lookup || !s.Factory.Factory.IPv6Lookup {
-		res, secondTrace, status, _ := s.DoMiekgLookup(miekg.Question{Name:name, Type:dns.TypeA}, nameServer)
+		res, secondTrace, status, _ := s.DoMiekgLookup(miekg.Question{Name: name, Type: dns.TypeA}, nameServer)
 		trace = append(trace, secondTrace...)
 		if status == zdns.STATUS_NOERROR {
 			cast, _ := res.(miekg.Result)
@@ -84,7 +84,7 @@ func (s *Lookup) LookupIPs(name, nameServer string) (CachedAddresses, zdns.Trace
 	}
 	// ipv6
 	if s.Factory.Factory.IPv6Lookup {
-		res, secondTrace, status, _ := s.DoMiekgLookup(miekg.Question{Name:name, Type:dns.TypeAAAA}, nameServer)
+		res, secondTrace, status, _ := s.DoMiekgLookup(miekg.Question{Name: name, Type: dns.TypeAAAA}, nameServer)
 		trace = append(trace, secondTrace...)
 		if status == zdns.STATUS_NOERROR {
 			cast, _ := res.(miekg.Result)
@@ -105,7 +105,7 @@ func (s *Lookup) LookupIPs(name, nameServer string) (CachedAddresses, zdns.Trace
 
 func (s *Lookup) DoLookup(name, nameServer string) (interface{}, zdns.Trace, zdns.Status, error) {
 	retv := Result{Servers: []MXRecord{}}
-	res, trace, status, err := s.DoMiekgLookup(miekg.Question{Name:name, Type: dns.TypeMX}, nameServer)
+	res, trace, status, err := s.DoMiekgLookup(miekg.Question{Name: name, Type: dns.TypeMX}, nameServer)
 	if status != zdns.STATUS_NOERROR {
 		return retv, trace, status, err
 	}

--- a/modules/nslookup/nslookup.go
+++ b/modules/nslookup/nslookup.go
@@ -50,7 +50,7 @@ func dotName(name string) string {
 
 func (s *Lookup) lookupIPs(name string, dnsType uint16, nameServer string) ([]string, zdns.Trace) {
 	var addresses []string
-	res, trace, status, _ := s.DoMiekgLookup(miekg.Question{Name:name, Type:dnsType}, nameServer)
+	res, trace, status, _ := s.DoMiekgLookup(miekg.Question{Name: name, Type: dnsType}, nameServer)
 	if status == zdns.STATUS_NOERROR {
 		if cast, ok := res.(miekg.Result); ok {
 			for _, innerRes := range cast.Answers {
@@ -65,7 +65,7 @@ func (s *Lookup) lookupIPs(name string, dnsType uint16, nameServer string) ([]st
 
 func (s *Lookup) DoNSLookup(name string, lookupIPv4, lookupIPv6 bool, nameServer string) (Result, zdns.Trace, zdns.Status, error) {
 	var retv Result
-	res, trace, status, err := s.DoMiekgLookup(miekg.Question{Name:name, Type:dns.TypeNS}, nameServer)
+	res, trace, status, err := s.DoMiekgLookup(miekg.Question{Name: name, Type: dns.TypeNS}, nameServer)
 	if status != zdns.STATUS_NOERROR || err != nil {
 		return retv, trace, status, nil
 	}

--- a/zdns/main.go
+++ b/zdns/main.go
@@ -224,9 +224,9 @@ func main() {
 	if gc.NameServerMode && gc.AlexaFormat {
 		log.Fatal("Alexa mode is incompatible with name server mode")
 	}
-        if gc.NameServerMode && gc.MetadataFormat {
-                log.Fatal("Metadata mode is incompatible with name server mode")
-        }
+	if gc.NameServerMode && gc.MetadataFormat {
+		log.Fatal("Metadata mode is incompatible with name server mode")
+	}
 	if gc.NameServerMode && gc.NameOverride == "" && gc.Module != "BINDVERSION" {
 		log.Fatal("Static Name must be defined with --override-name in --name-server-mode unless DNS module does not expect names (e.g., BINDVERSION).")
 	}

